File: app/src/main/java/org/akvo/mapexperiments/MapBoxActivity.java
package org.akvo.mapexperiments;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.mapbox.mapboxsdk.annotations.Icon;
import com.mapbox.mapboxsdk.annotations.IconFactory;
import com.mapbox.mapboxsdk.annotations.MarkerViewOptions;
import com.mapbox.mapboxsdk.annotations.PolylineOptions;
import com.mapbox.mapboxsdk.camera.CameraPosition;
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.geometry.LatLngBounds;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.offline.OfflineRegion;
import com.mapbox.mapboxsdk.offline.OfflineRegionDefinition;
import com.mapbox.mapboxsdk.offline.OfflineTilePyramidRegionDefinition;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class MapBoxActivity extends LocationAwareActivity
        implements RegionsListDialogFragment.RegionsSelectionListener {

    public static final String PREFERENCE_NAME = "Prefs";
    public static final String PREF_SHAPE = "shape";
    public static final double ZOOM_LEVEL = 16;

    private final BitmapGenerator bitmapGenerator = new BitmapGenerator();
    private final List<LatLng> locations = new ArrayList<>();
    private final Gson gson = new Gson();

    private MapboxMap mapboxMap;
    private MapView mapView;
    private SharedPreferences sharedPreferences;
    private boolean manualAreaSelected;
    private Icon icon;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_mapbox);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        sharedPreferences = getSharedPreferences(PREFERENCE_NAME, MODE_PRIVATE);

        IconFactory iconFactory = IconFactory.getInstance(MapBoxActivity.this);
        Bitmap bmp = bitmapGenerator.getBitmap();
        icon = iconFactory.fromBitmap(bmp);

        mapView = findViewById(R.id.mapView);
        mapView.onCreate(savedInstanceState);
        mapView.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(MapboxMap mapboxMap) {
                MapBoxActivity.this.mapboxMap = mapboxMap;
                checkLocation();
            }
        });
    }

    @Override
    protected void updateNewLocation(@Nullable Location location) {
        if (location == null) {
            return;
        }
        LatLngAcc latestLatLng = new LatLngAcc(location);

        LatLng lastLatLong = locations.size() == 0 ? null : locations.get(locations.size() - 1);
        //1 meters minimum distance for the point to be added
        if (lastLatLong == null
                || latestLatLng.distanceTo(lastLatLong) > MapOptions.MINIMUM_DISTANCE) {
            if (!manualAreaSelected) {
                double zoom = lastLatLong == null ? ZOOM_LEVEL : mapboxMap.getCameraPosition().zoom;
                if (zoom == 0) {
                    zoom = ZOOM_LEVEL;
                }
                mapboxMap.animateCamera(CameraUpdateFactory.newLatLngZoom(
                        new LatLng(location.getLatitude(), location.getLongitude()), zoom));
            }
            locations.add(latestLatLng);
            redrawMap();
        }
    }

    @Override
    protected void enableMapLocation() {
        //Ignore
    }

    private void redrawMap() {
        mapboxMap.clear();

        //TODO: save geolines to json each time

        mapboxMap.addPolyline(new PolylineOptions()
                .addAll(locations)
                .color(MapOptions.LINE_COLOR)
                .width(MapOptions.LINE_WIDHT));
        for (LatLng latLng : locations) {
            mapboxMap.addMarker(new MarkerViewOptions()
                    .anchor(MapOptions.MARKER_ANCHOR, MapOptions.MARKER_ANCHOR)
                    .icon(icon)
                    .position(latLng)
                    .title(((LatLngAcc) latLng).getTitle()));
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        mapView.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mapView.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mapView.onPause();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mapView.onStop();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        mapView.onSaveInstanceState(outState);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.mapbox_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.save:
                if (locations.size() > 0) {
                    sharedPreferences.edit()
                            .putString(PREF_SHAPE, gson.toJson(locations)).apply();
                    toast("Shape saved");
                }
                return true;
            case R.id.load:
                String savedLocationsString = sharedPreferences.getString(PREF_SHAPE, null);
                if (!TextUtils.isEmpty(savedLocationsString)) {
                    toast("Loading shape");
                    Type type = new TypeToken<List<LatLngAcc>>() {
                    }.getType();
                    List<LatLngAcc> locationsPref = gson.fromJson(savedLocationsString, type);
                    locations.clear();
                    locations.addAll(locationsPref);
                    //TODO: center map on last location item.
                    redrawMap();
                }
                return true;
            case R.id.load_offline:
                RegionsListDialogFragment fragment = RegionsListDialogFragment.newInstance();
                fragment.show(getSupportFragmentManager(), RegionsListDialogFragment.TAG);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void toast(String text) {
        Toast.makeText(this, text, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onAddNewSelected() {
        startActivity(new Intent(MapBoxActivity.this, OfflineManagerActivity.class));
    }

    @Override
    public void onRegionSelected(OfflineRegion region) {
        OfflineRegionDefinition definition = region.getDefinition();
        LatLngBounds bounds = definition
                .getBounds();
        double regionZoom = ((OfflineTilePyramidRegionDefinition)
                definition)
                .getMinZoom();

        // Create new camera position
        CameraPosition cameraPosition = new CameraPosition.Builder()
                .target(bounds.getCenter())
                .zoom(regionZoom)
                .build();

        // Move camera to new position
        mapboxMap.moveCamera(CameraUpdateFactory
                .newCameraPosition(cameraPosition));
        manualAreaSelected = true;
    }

    public class LatLngAcc extends LatLng {

        private float accuracy = 0.0f;

        LatLngAcc(Location location) {
            super(location);
            this.accuracy = location.getAccuracy();
        }

        public float getAccuracy() {
            return accuracy;
        }

        public String getTitle() {
            return "Loc: " + getLatitude() + ", " + getLongitude()
                    + ", acc: " + accuracy;
        }
    }
}

File: app/src/main/java/org/akvo/mapexperiments/MapBoxActivity.kt
package org.akvo.mapexperiments

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.location.Location
import android.os.Bundle
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.mapbox.mapboxsdk.camera.CameraPosition
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory
import com.mapbox.mapboxsdk.geometry.LatLng
import com.mapbox.mapboxsdk.maps.MapView
import com.mapbox.mapboxsdk.maps.MapboxMap
import com.mapbox.mapboxsdk.maps.Style
import com.mapbox.mapboxsdk.offline.OfflineRegion
import com.mapbox.mapboxsdk.offline.OfflineTilePyramidRegionDefinition
import com.mapbox.mapboxsdk.plugins.annotation.CircleManager
import com.mapbox.mapboxsdk.plugins.annotation.CircleOptions
import com.mapbox.mapboxsdk.plugins.annotation.FillManager
import com.mapbox.mapboxsdk.plugins.annotation.FillOptions
import com.mapbox.mapboxsdk.plugins.annotation.LineManager
import com.mapbox.mapboxsdk.plugins.annotation.LineOptions
import org.akvo.mapexperiments.RegionsListDialogFragment.RegionsSelectionListener
import java.util.ArrayList

class MapBoxActivity : LocationAwareActivity(), RegionsSelectionListener {
    private val locations: MutableList<LatLng> = ArrayList()
    private val gson = Gson()
    private lateinit var mapboxMap: MapboxMap
    private lateinit var mapView: MapView
    private var sharedPreferences: SharedPreferences? = null
    private var manualAreaSelected = false
    private var enableTracking = false

    private lateinit var circleManager: CircleManager
    private lateinit var lineManager: LineManager
    private lateinit var fillManager: FillManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_mapbox)
        val toolbar =
            findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        sharedPreferences = getSharedPreferences(
            PREFERENCE_NAME,
            Context.MODE_PRIVATE
        )
        mapView = findViewById(R.id.mapView)
        mapView.onCreate(savedInstanceState)
        mapView.getMapAsync { mapboxMap ->
            onMapReady(mapboxMap)
        }
        findViewById<Button>(R.id.button2).setOnClickListener { v -> onButtonClick(v as Button) }
    }

    private fun onButtonClick(v: Button) {
        if (enableTracking) {
            v.setText(R.string.start_recording)
            enableTracking = false
            saveLocations()
            findViewById<TextView>(R.id.textView2).text = ""
        } else {
            locations.clear()
            redrawMap()
            enableTracking = true
            v.setText(R.string.stop_recording)
        }
    }

    private fun onMapReady(mapboxMap: MapboxMap) {
        this@MapBoxActivity.mapboxMap = mapboxMap
        mapboxMap.setStyle(Style.LIGHT) { style ->
            lineManager = LineManager(mapView, mapboxMap, style)
            circleManager = CircleManager(mapView, mapboxMap, style)
            fillManager = FillManager(mapView, mapboxMap, style)
            checkLocation()
        }

    }

    override fun updateNewLocation(location: Location?) {
        if (enableTracking) {
            if (location == null) {
                return
            }
            val latestLatLng = LatLngAcc(location)
            findViewById<TextView>(R.id.textView2).text = getString(R.string.last_location, latestLatLng.title)
            val lastLatLong =
                if (locations.size == 0) null else locations[locations.size - 1]
            //5 meters minimum distance for the point to be added
            if (lastLatLong == null
                || latestLatLng.distanceTo(lastLatLong) > MapOptions.MINIMUM_DISTANCE
            ) {
                if (!manualAreaSelected) {
                    var zoom =
                        if (lastLatLong == null) ZOOM_LEVEL else mapboxMap.cameraPosition.zoom
                    if (zoom == 0.0) {
                        zoom = ZOOM_LEVEL
                    }
                    mapboxMap.animateCamera(
                        CameraUpdateFactory.newLatLngZoom(
                            LatLng(location.latitude, location.longitude),
                            zoom
                        )
                    )
                }
                locations.add(latestLatLng)
                redrawMap()
            }
        }
    }

    override fun enableMapLocation() {
        //Ignore
    }

    private fun redrawMap() {
        circleManager.deleteAll()
        lineManager.deleteAll()
        fillManager.deleteAll()

        fillManager.create(
            FillOptions()
                .withLatLngs(arrayListOf(locations))
                .withFillColor("#736357")
                .withFillOpacity(0.5f)
                .withDraggable(false)
        )

        lineManager.create(
            LineOptions()
                .withLineColor("#736357")
                .withLineWidth(4f)
                .withDraggable(false)
                .withLatLngs(locations)
        )

        for (latLng in locations) {
            circleManager.create(
                CircleOptions()
                    .withLatLng(latLng)
                    .withCircleColor("#00a79d")
                    .withCircleRadius(8f)
                    .withCircleStrokeColor("#027a73")
                    .withDraggable(false)
                    .withCircleStrokeWidth(1f)
            )
        }
    }

    override fun onStart() {
        super.onStart()
        mapView.onStart()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onStop() {
        super.onStop()
        mapView.onStop()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mapView.onSaveInstanceState(outState)
    }

    override fun onDestroy() {
        super.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView.onLowMemory()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.mapbox_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.save -> {
                saveLocations()
                true
            }
            R.id.load -> {
                val savedLocationsString =
                    sharedPreferences!!.getString(PREF_SHAPE, null)
                if (!TextUtils.isEmpty(savedLocationsString)) {
                    toast("Loading shape")
                    val type =
                        object : TypeToken<List<LatLngAcc?>?>() {}.type
                    val locationsPref =
                        gson.fromJson<List<LatLngAcc>>(
                            savedLocationsString,
                            type
                        )
                    locations.clear()
                    locations.addAll(locationsPref)
                    //TODO: center map on last location item.
                    redrawMap()
                }
                true
            }
            R.id.load_offline -> {
                val fragment = RegionsListDialogFragment.newInstance()
                fragment.show(supportFragmentManager, RegionsListDialogFragment.TAG)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun saveLocations() {
        val size = locations.size
        if (size > 0) {
            if (size > 2) {
                locations.add(locations[0])
            }
            val toJson = gson.toJson(locations)
            sharedPreferences!!.edit()
                .putString(PREF_SHAPE, toJson)
                .apply()
            toast("Shape saved")
            redrawMap()
        }
    }

    private fun toast(text: String) {
        Toast.makeText(this, text, Toast.LENGTH_LONG).show()
    }

    override fun onAddNewSelected() {
        startActivity(Intent(this@MapBoxActivity, OfflineManagerActivity::class.java))
    }

    override fun onRegionSelected(region: OfflineRegion) {
        val definition = region.definition
        val bounds = definition
            .bounds
        val regionZoom = (definition as OfflineTilePyramidRegionDefinition)
            .minZoom

        // Create new camera position
        val cameraPosition =
            CameraPosition.Builder()
                .target(bounds.center)
                .zoom(regionZoom)
                .build()

        // Move camera to new position
        mapboxMap.moveCamera(
            CameraUpdateFactory
                .newCameraPosition(cameraPosition)
        )
        manualAreaSelected = true
    }

    class LatLngAcc internal constructor(location: Location) :
        LatLng(location) {
        var accuracy = 0.0f

        val title: String
            get() = ("Loc: " + latitude + ", " + longitude
                    + ", acc: " + accuracy)

        init {
            accuracy = location.accuracy
        }
    }

    companion object {
        const val PREFERENCE_NAME = "Prefs"
        const val PREF_SHAPE = "shape"
        const val ZOOM_LEVEL = 16.0
    }
}